posts/2024-05-01-spring-boot-3.adoc

```diff
@@ -0,0 +1,239 @@
+
+If you're wondering, yes, while a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment, that executable WAR could still be deployed as a standard WAR.
+
+=== Part 3: Build container image with efficient layering
+
+Now that we have used the optimized Spring Boot deployment, we can efficiently build a container image.  This image uses an indexed cache at `/lib.index.cache` to store Spring Boot dependencies in their own layer, separate from your application code.
+
+1. Create your `Dockerfile`.  
++
+[source,dockerfile]
+----
+# Stage and thin the application 
+FROM icr.io/appcafe/open-liberty:full-java21-openj9-ubi-minimal as staging
+
+ARG APPNAME=spring-liberty-0.0.1-SNAPSHOT.war
+COPY --chown=1001:0 target/$APPNAME \
+  /staging/$APPNAME
+
+RUN springBootUtility thin \
+ --sourceAppPath=/staging/$APPNAME \
+ --targetThinAppPath=/staging/thin-$APPNAME \
+ --targetLibCachePath=/staging/lib.index.cache
+
+FROM icr.io/appcafe/open-liberty:kernel-slim-java21-openj9-ubi-minimal
+
+ARG APPNAME=spring-liberty-0.0.1-SNAPSHOT.war
+ARG VERSION=1.0
+ARG REVISION=SNAPSHOT
+COPY --chown=1001:0 src/main/liberty/config/server.xml /config/server.xml
+
+RUN features.sh
+
+COPY --chown=1001:0 --from=staging /staging/lib.index.cache /lib.index.cache
+COPY --chown=1001:0 --from=staging /staging/thin-$APPNAME \
+                    /config/apps/thin-$APPNAME
+
+RUN configure.sh 
+----
++
+(Note you can use the `full-java17-openj9-ubi` tag to build the equivalent Java 17 image.)
++
+2. Build then run the image.
++
+[source,sh]
+----
+docker build -t springboot:demo .
+docker run -p 9080:9080 -p 9443:9443 -it springboot:demo
+----
+
+== Results
+
+To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can configure an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature in the `server.xml` file.
+Though much of the programming model is the same across the two cases, there are some differences, including the bootstrap mechanism and the ability to create more efficient Docker layers with the optimized deployment.
+
+(**Note:** the flow through Parts 1 and 2 of this blog post are a bit contrived to show the two deployment options, and your real development workflow would probably look a bit different. When using a traditional WAR deployment you would likely maintain your own `server.xml` file rather than continuing to use the template configuration. When using the optimized Liberty deployment there's no need to do a traditional deployment first.)
```

Comment:
```suggestion
**Note:** The flow through Parts 1 and 2 of this blog post is a bit contrived to show the two deployment options, and your real development workflow would probably look a bit different. When using a traditional WAR deployment, you would likely maintain your own `server.xml` file rather than continuing to use the template configuration. When using the optimized Liberty deployment, there's no need to do a traditional deployment first.
```